### ISSUE TITLE
memory leakage caused by logging_output in criterion "masked_lm"

### ISSUE THREAD

**Issue by hury07** [pytorch/fairseq#2255]
## 🐛 Bug
I found memory (not GPU memory) leakage event when trying masked language modeling task.
The memory usage was increasing when the trainning proceeded.

The bug was caused by 'loss' item in logging_output in the file criterions/masked_lm.py.
The original code is,
 64         logging_output = {
 65             'loss': loss,
 66             'ntokens': sample['ntokens'],
 67             'nsentences': sample['nsentences'],
 68             'sample_size': sample_size,
 69         }

The bug could be fixed with the following code.
 64         logging_output = {
 65             'loss': loss.data,
 66             'ntokens': sample['ntokens'],
 67             'nsentences': sample['nsentences'],
 68             'sample_size': sample_size,
 69         }

**Comment by myleott** [comment]
Interesting, can you confirm that using `.detach()` instead of `.data` works as well? I believe `detach()` is the preferred way of doing this now.

**Comment by myleott** [comment]
@joshim5 has a fix for this

**Comment by joshim5** [comment]
Landed in https://github.com/pytorch/fairseq/commit/8eb9123f560d32940f96a01369d61c1684dce085. @hury07 please reopen this issue if you still encounter problems.
